include/sparrow/arrow_array_schema_proxy.hpp

```diff
@@ -95,100 +132,180 @@ namespace sparrow
     }
 
-    inline
-    arrow_proxy::arrow_proxy(ArrowArray&& array, ArrowSchema* schema)
-        : m_array(std::move(array))
-        , m_schema(schema)
+    inline arrow_proxy::arrow_proxy(ArrowArray&& array, ArrowSchema&& schema)
+        : arrow_proxy(std::move(array), std::move(schema), true)
+    {
+    }
+
+    inline arrow_proxy::arrow_proxy(ArrowArray&& array, ArrowSchema* schema)
+        : arrow_proxy(std::move(array), schema, true)
+    {
+    }
+
+    inline arrow_proxy::arrow_proxy(ArrowArray* array, ArrowSchema* schema)
+        : arrow_proxy(array, schema, true)
+    {
+    }
+
+    inline arrow_proxy::arrow_proxy(const arrow_proxy& other)
+        : m_array(deep_copy_array(other.m_array_ref, other.m_schema_ref))
+        , m_array_ref(std::get<1>(m_array))
+        , m_schema(deep_copy_schema(other.m_schema_ref))
+        , m_schema_ref(std::get<1>(m_schema))
     {
-        SPARROW_ASSERT_TRUE(schema != nullptr);
         validate_array_and_schema();
-        initialize_children();
         initialize_buffers();
     }
 
-    inline
-    arrow_proxy::arrow_proxy(ArrowArray* array, ArrowSchema* schema)
-        : m_array(array)
-        , m_schema(schema)
+    inline arrow_proxy& arrow_proxy::operator=(const arrow_proxy& other)
     {
-        SPARROW_ASSERT_TRUE(array != nullptr);
-        SPARROW_ASSERT_TRUE(schema != nullptr);
+        if (this == &other)
+        {
+            return *this;
+        }
+
+        m_array = ArrowArray{};
+        m_array_ref = std::get<1>(m_array);
+        deep_copy_array(other.m_array_ref, other.m_schema_ref, m_array_ref);
+
+        m_schema = ArrowSchema{};
+        m_schema_ref = std::get<1>(m_schema);
+        deep_copy_schema(other.m_schema_ref, m_schema_ref);
+
         validate_array_and_schema();
-        initialize_children();
         initialize_buffers();
+
+        return *this;
     }
 
-    inline
-    arrow_proxy::~arrow_proxy()
+    inline arrow_proxy::arrow_proxy(arrow_proxy&& other)
+        : m_array(std::move(other.m_array))
+        , m_array_ref(m_array.index() == 0 ? *std::get<0>(m_array) : std::get<1>(m_array))
+        , m_schema(std::move(other.m_schema))
+        , m_schema_ref(m_schema.index() == 0 ? *std::get<0>(m_schema) : std::get<1>(m_schema))
+    {
+        other.m_array = {};
+        other.m_schema = {};
+        other.m_buffers.clear();
+
+        initialize_buffers();
+    }
+
+    inline arrow_proxy& arrow_proxy::operator=(arrow_proxy&& rhs)
     {
-        if (m_array.index() == 1)
+        if (this == &rhs)
         {
-            ArrowArray& array = std::get<1>(m_array);
-            if (array.release != nullptr)
+            return *this;
+        }
+
+        m_array = std::move(rhs.m_array);
+        m_array_ref = m_array.index() == 0 ? *std::get<0>(m_array) : std::get<1>(m_array);
+        m_schema = std::move(rhs.m_schema);
+        m_schema_ref = m_schema.index() == 0 ? *std::get<0>(m_schema) : std::get<1>(m_schema);
```

Comment:
This implementation may leak as `m_array` and `m_schema` are never released while they should. A elegant solution is to swap `m_array` and `rhs.m_array` (same for `schema` and `rhs.m_schema`). Notice that `m_array_ref` and `m_schema_ref` should be pointers and handled the same way. Here the object referenced by `m_schema_ref` is assigned a copy of the content of `m_schema`, which is probably not what you want.

Reply:
done